### ISSUE TITLE
aws_s3 sink - user is not authorized to perform sts:AssumeRole on resource role

### ISSUE THREAD

**Issue by xavigpich** [vectordotdev/vector#11114]
### Vector Version
```
vector 0.19.0 (x86_64-unknown-linux-gnu da60b55 2021-12-28)
```

### Vector Configuration File

```toml  
[sources.beats]
  type    = "file"
  include = [ "/var/log/auditbeat/*.json" ]

[sinks.aws_s3]
  type             = "aws_s3"
  inputs           = [ "beats" ]
  bucket           = "<my-bucket>"
  region           = "<my-region>"
  auth.assume_role = "arn:aws:iam::<my-account-id>:role/<my-role>"
  key_prefix       = "logs/beats/year=%Y/month=%m/day=%d/"
  compression      = "gzip"
  encoding.codec   = "ndjson"
```

### Debug Output
```
RUST_BACKTRACE=full vector -vvv --config=/etc/vector/vector.toml
...
User: arn:aws:sts::<my-account-id>:assumed-role/<my-role>/<my-instance-id> is not authorized to perform: 
sts:AssumeRole on resource: arn:aws:iam::<my-account-id>:role/<my-role>
```
Please, see the full debug output in the GitHub gist [file](https://gist.github.com/thisisxgp/3d06cc567e3f028da2365d07ca357133)

### Expected Behavior
Logs should be written into AWS S3 bucket in the specified path

### Actual Behavior
Vector fails at healthcheck and gets `AccessDenied` on the bucket using the assumed role 
```
2022-01-31T19:07:05.117380Z ERROR vector::topology::builder: msg="Healthcheck: Failed Reason." error=
StsProvider get_session_token error: Unknown(BufferedHttpResponse {status: 403, body: "<ErrorResponse 
xmlns=\"https://sts.amazonaws.com/doc/2011-06-15/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>AccessDenied</Code>\n    
<Message>User: arn:aws:sts::<my-account-id>:assumed-role/<my-role>/<my-instance-id> is not authorized to perform: 
sts:AssumeRole on resource: arn:aws:iam::<my-account-id>:role/<my-role></Message>\n  </Error>\n  
<RequestId>f<my-request-id></RequestId>\n</ErrorResponse>\n", headers: {"x-amzn-requestid": <my-request-id>, 
"content-type": "text/xml", "content-length": "471", "date": "Mon, 31 Jan 2022 19:07:04 GMT"} }) component_kind="sink" 
component_type="aws_s3" component_id=aws_s3 [component_name=aws_s3]
```

### Additional Context
Vector is installed in a AWS EC2 instance. From the instance (as root user), we are able to read/write to the AWS S3 bucket. We've double checked and `s3:PutObject` and `s3:HeadBucket` permissions are set as per the [docs](https://vector.dev/docs/reference/configuration/sinks/aws_s3/#permissions). From the command line, we are able to assume the role:
```
aws sts assume-role --role-arn arn:aws:iam::<my-account-id>:role/<my-role> --role-session-name <my-instance-id>
{
    "AssumedRoleUser": {
        "AssumedRoleId": <ASSUMED_ROLE_ID>, 
        "Arn": "arn:aws:sts::<my-account-id>:assumed-role/<my-role>/<my-instance-id>"
    }, 
    "Credentials": {
        "SecretAccessKey": <SECRET_ACCESS_KEY>, 
        "SessionToken": <SESSION_TOKEN>, 
        "Expiration": "2022-01-31T19:25:16Z", 
        "AccessKeyId": <ACCESS_KEY_ID>
    }
}
```
We've used these credentials in the `.toml` configuration file and got the same errors.
We've also tried `[transforms.aws_ec2_metadata]` and got the same errors.
We'd prefer not to use `auth` credentials as they are temporarily rotated every few minutes in the instance. 

Please, can someone take a look into it? Hopefully, this is familiar to someone.
Thanks!
-X

**Comment by jszwedko** [comment]
Closing due to inactivity, but feel free to re-open if you come back around to this. Apologies again for the initial delay in response.

**Comment by jszwedko** [comment]
I suspect this also may have been fixed in 0.21 when we switched to the official AWS Rust SDK.

**Comment by betabrandao** [comment]
One thing I noticed in version 0.28.0, is that the error still persists. But there is a notation in the error that went unnoticed:

In AWS Doc:
**Amazon EC2 does not have permission to assume the role**. Permission to assume the role is controlled by the trust policy attached to the role, like the example that follows. Use the IAM UpdateAssumeRolePolicy API to update the trust policy.
https://docs.aws.amazon.com/IAM/latest/UserGuide/troubleshoot_iam-ec2.html#troubleshoot_iam-ec2_errors-info-doc

So, the action of assuming the role of an instance will not work. It is a limitation of using an EC2 Instance. I believe this can work inside a container, but I haven't tested this possibility.

My workaround:I created a cron that updates the AWS credentials by querying the EC2 api:
export AWS_ACCESS_KEY_ID=$(curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/<ROLE_NAME> | awk '{print $3}' | xargs | cut -d',' -f4 | sed 's/ //g')
export AWS_SECRET_ACCESS_KEY=$(curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/<ROLE_NAME> | awk '{print $3}' | xargs | cut -d',' -f5 | sed 's/ //g')
export AWS_SESSION_TOKEN=$(curl -s http://169.254.169.254/latest/meta-data/iam/security-credentials/<ROLE_NAME> | awk '{print $3}' | xargs | cut -d',' -f6| sed 's/ //g')
```bash

```
